File: Java/src/_0203_Remove_Linked_List_Elements/Solution.java
package _0203_Remove_Linked_List_Elements;

public class Solution {

    public ListNode removeElements(ListNode head, int val) {

        while (head != null && head.val == val) {
            ListNode delNode = head;
            head = head.next;
            delNode.next = null;
        }
        if (head == null) return head;

        ListNode preNode = head;
        while (preNode.next != null) {
            if (preNode.next.val == val) {
                ListNode delNode = preNode.next;
                preNode.next = delNode.next;
                delNode.next = null;
            } else {
                preNode = preNode.next;
            }
        }

        return head;
    }

    public static void main(String[] args) {

        int[] nums = {1, 2, 6, 3, 4, 5, 6};
        // 创建链表
        MyLinkedList myLinkedList = new MyLinkedList(nums);
        ListNode head = myLinkedList.getHead();
        System.out.println(myLinkedList);
        // 删除元素
        (new Solution()).removeElements(head, 6);
        // 输出链表
        System.out.println(myLinkedList);
    }
}

File: Python/_0203_Remove_Linked_List_Elements/Solution.py
# coding=utf-8
# Date: 2021/6/5 14:06
from MyLinkedList import *


# 执行用时：76 ms, 在所有 Python3 提交中击败了44.93%的用户
# 内存消耗：25.9 MB, 在所有 Python3 提交中击败了5.02%的用户
class Solution:
    def removeElements(self, head: ListNode, val: int) -> ListNode:
        if not head:
            return None
        head.next = self.removeElements(head.next, val)
        return head.next if head.val == val else head


if __name__ == "__main__":
    head = MyLinkedList().createLinkedList([1, 2, 3, 4, 5, 6])
    MyLinkedList().printLinkedList(Solution().removeElements(head, 6))
